src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatDefaultPayload.cs

```diff
@@ -2,46 +2,251 @@
                             break;
                         case "baseSdkTargetFramework":
-                            provider.AddHeartbeatProperty(fieldName, GetBaseSdkTargetFramework(), true);
+                            provider.AddHeartbeatProperty(fieldName, true, HeartbeatDefaultPayload.GetBaseSdkTargetFramework(), true);
+                            break;
+                        case "osVersion":
+                            provider.AddHeartbeatProperty(fieldName, true, HeartbeatDefaultPayload.GetRuntimeOsType(), true);
+                            break;
+                        case "processSessionId":
+                            provider.AddHeartbeatProperty(fieldName, true, HeartbeatDefaultPayload.GetProcessSessionId(), true);
+                            break;
+                        case "osType":
+                        case "location":
+                        case "name":
+                        case "offer":
+                        case "platformFaultDomain":
+                        case "platformUpdateDomain":
+                        case "publisher":
+                        case "sku":
+                        case "version":
+                        case "vmId":
+                        case "vmSize":
+                            // skip all Azure Instance Metadata fields
                             break;
                         default:
-                            provider.AddHeartbeatProperty(fieldName, "UNDEFINED", false);
+                            provider.AddHeartbeatProperty(fieldName, true, "UNDEFINED", false);
                             break;
                     }
                 }
                 catch (Exception ex)
                 {
                     CoreEventSource.Log.FailedToObtainDefaultHeartbeatProperty(fieldName, ex.ToString());
                 }
+                finally
+                {
+                    SdkInternalOperationsMonitor.Exit();
+                }
             }
         }
 
+        /// <summary>
+        /// Gathers operational data from an Azure VM, if the IMDS endpoint is present. If not, and the SDK is
+        /// not running on an Azure VM with the Azure Instance Metadata Service running, the corresponding
+        /// fields will not get set into the heartbeat payload.
+        /// </summary>
+        private static async Task AddAzureVmDetail(HeartbeatProvider heartbeatManager, IEnumerable<string> enabledFields)
+        {
+            if (HeartbeatDefaultPayload.EnableAzureInstanceMetadata && !HeartbeatDefaultPayload.isAzureMetadataCheckCompleted)
+            {
+                // only ever do this once when the SDK gets initialized
+                HeartbeatDefaultPayload.isAzureMetadataCheckCompleted = true;
+
+                var allFields = await GetAzureInstanceMetadataFields(baseImdsUrl, imdsApiVersion, imdsTextFormat)
```

Comment:
Can we determine that we're running on Azure infra before making this call? Concern - we will be doing this call for On-Premise VMs and other APM tools will be flagging this call.